Tests/tests/test_TraceControl.cpp

```diff
@@ -0,0 +1,420 @@
+const string webPrefix = _T("/Service/TraceControl");
+const string volatilePath = _T("/tmp/");
+}
+
+class TraceControlTest : public ::testing::Test {
+protected:
+    Core::ProxyType<Plugin::TraceControl> plugin;
+    PluginHost::IWeb* interface;
+    Core::JSONRPC::Handler& handler;
+    Core::JSONRPC::Connection connection;
+    string response;
+
+    TraceControlTest()
+        : plugin(Core::ProxyType<Plugin::TraceControl>::Create())
+        , handler(*(plugin))
+        , connection(1, 0)
+    {
+        interface = static_cast<PluginHost::IWeb*>(plugin->QueryInterface(PluginHost::IWeb::ID));
+        Trace::TraceUnit::Instance().Open(volatilePath);
+    }
+
+    virtual ~TraceControlTest()
+    {
+        interface->Release();
+        plugin.Release();
+        Trace::TraceUnit::Instance().Close();
+    }
+};
+
+class TraceControlInitializedTest : public TraceControlTest {
+protected:
+    FactoriesImplementation factoriesImplementation;
+    IarmBusImplMock iarmBusImplMock;
+    ServiceMock service;
+    COMLinkMock comLinkMock;
+
+    TraceControlInitializedTest()
+        : TraceControlTest()
+    {
+    }
+    virtual ~TraceControlInitializedTest() override
+    {
+    }
+
+    void SetUp(bool bBackground)
+    {
+        ON_CALL(service, Background())
+            .WillByDefault(::testing::Return(bBackground));
+        ON_CALL(service, ConfigLine())
+            .WillByDefault(::testing::Return("{}"));
+        ON_CALL(service, WebPrefix())
+            .WillByDefault(::testing::Return(webPrefix));
+        ON_CALL(service, VolatilePath())
+            .WillByDefault(::testing::Return(volatilePath));
+        ON_CALL(service, Callsign())
+            .WillByDefault(::testing::Return(callSign));
+         ON_CALL(service, COMLink())
+             .WillByDefault(::testing::Return(&comLinkMock));
+
+        PluginHost::IFactories::Assign(&factoriesImplementation);
+        IarmBus::getInstance().impl = &iarmBusImplMock;
+        EXPECT_EQ(string(""), plugin->Initialize(&service));
+    }
+
+    virtual void TearDown()
+    {
+        plugin->Deinitialize(&service);
+        PluginHost::IFactories::Assign(nullptr);
+        IarmBus::getInstance().impl = nullptr;
+    }
+};
+
+TEST_F(TraceControlInitializedTest, registeredMethods)
+{
+    SetUp(true);
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("set")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("status")));
+}
+
+TEST_F(TraceControlInitializedTest, jsonRpc)
+{
+    SetUp(true);
+    EXPECT_EQ(Core::ERROR_NONE, handler.Invoke(connection, _T("status"), _T("{}"), response));
+
+    JsonObject params;
+    params.FromString(response);
+    JsonArray arr = params["settings"].Array();
+    JsonObject traceObj;
+    string jsonString;
+    for (unsigned int i = 0; i < arr.Length(); i++)
+    {
+        traceObj = arr[i].Object();
+        traceObj.ToString(jsonString);
+        EXPECT_THAT(jsonString, ::testing::MatchesRegex(_T("\\{"
```

Comment:
please make array regex instead

Reply:
The problem with using array regex is the array can be of different size based on what modules are enabled/disabled.
Here is the example json output. If one plugin is not available then it may not show in this list and same time in future if a new plugin gets added that test might fail with the exact match check.
I added a generic check that will take care of these cases. The same comment will be applied to other remarks in this review.

{"settings":[{"module":"Plugin_DeviceIdentification","category":"Information","state":"disabled"},{"module":"Plugin_DeviceIdentification","category":"Error","state":"disabled"},{"module":"Plugin_DeviceInfo","category":"Fatal","state":"disabled"},{"module":"Plugin_DeviceInfo","category":"Information","state":"disabled"},{"module":"Plugin_LocationSync","category":"Information","state":"disabled"},{"module":"Plugin_LocationSync","category":"Warning","state":"disabled"},{"module":"Plugin_LocationSync","category":"Error","state":"disabled"},{"module":"Plugin_Messenger","category":"Warning","state":"disabled"},{"module":"Plugin_Messenger","category":"Information","state":"disabled"},{"module":"Plugin_Messenger","category":"Error","state":"disabled"},{"module":"Plugin_SystemServices","category":"Error","state":"disabled"},{"module":"Plugin_SystemServices","category":"Information","state":"disabled"},{"module":"Plugin_TimeSync","category":"Information","state":"disabled"},{"module":"Plugin_TraceControl","category":"Information","state":"disabled"},{"module":"Plugin_TraceControl","category":"Error","state":"disabled"},{"module":"SysLog","category":"ParsingError","state":"enabled"},{"module":"SysLog","category":"Crash","state":"enabled"},{"module":"SysLog","category":"Startup","state":"enabled"},{"module":"SysLog","category":"Shutdown","state":"enabled"},{"module":"SysLog","category":"Notification","state":"enabled"},{"module":"SysLog","category":"Error","state":"enabled"},{"module":"SysLog","category":"Fatal","state":"enabled"}]}

Reply:
it should not be a problem to write a regex
you can check how it's done in other places e.g. https://github.com/rdkcentral/rdkservices/blob/dcaf5c0267ed9e645b70825095252eb13a53dbe8/Tests/tests/test_DeviceInfoWeb.cpp#L103

Reply:
I was able to use the regex from https://github.com/rdkcentral/rdkservices/pull/3345 and update the code.
Some conditional regex are left as it is by parsing the array.

Reply:
please remove conditional regex as well. you can split it in 2 regex if you like, 1 to verify response, another to verify specific part of it. if-else is not good, as it doesn't check the specific part is present

Reply:
Can you let me know, which specific part the if-else is not checking?

Reply:
it is not checking that Plugin_TraceControl is present. It is sad that a test uses if-else like it's random response when it is not

Reply:
Thanks for the finding. I added the check for this module existence. In this particular case I am only interested in "Plugin_TraceControl" module and the rest of the string can be random. I am not expert in regex, so instead of creating complicated regex and test those, I chose to add these conditions.

Reply:
it needs to be reworked without if-else. if you struggle with that I will have to address it after your change

Reply:
I am fine with it.

Reply:
https://ccp.sys.comcast.net/browse/RDKCOM-3396